tests/tests_eve/unit_tests/test_datamodels.py

```diff
@@ -0,0 +1,892 @@
+# -*- coding: utf-8 -*-
+#
+# Eve Toolchain - GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2020, CSCS - Swiss National Supercomputing Center, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+
+from __future__ import annotations
+
+import dataclasses
+import enum
+import inspect
+import random
+import types
+import typing
+from typing import (
+    Any,
+    Callable,
+    ClassVar,
+    Dict,
+    Generic,
+    List,
+    Literal,
+    Mapping,
+    MutableSequence,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    TypeVar,
+    Union,
+)
+
+import factory
+import pytest
+import pytest_factoryboy as pytfboy
+
+from eve import datamodels, utils
+
+
+T = TypeVar("T")
+
+model_factories: List[factory.Factory] = []
+invalid_model_factories: List[factory.Factory] = []
+
+
+def register_factory(
+    factory_or_name: Optional[Union[factory.Factory, str]] = None,
```

Comment:
In this way it will work if it is used either as `@register_factory` or `@register_factory()`

Reply:
I guess I don't understand how decorators work. Can there be a case in which it is None? And if yes why does it make sense?

Reply:
Remember that
```
@decorator
def f():
    ...
```
It's just syntatic sugar for: ```f = decorator(f)```

And then:
```
def decorator(arg=None):
    assert callable(arg)

# this will fail because the decorator receives None: function = decorator()(function)
@decorator()
def function():
    pass

# this will work as intended: function = decorator(function)
@decorator
def function():
    pass
```
Since it's very easy to forget about the parenthesis, specially if the decorator accepts other optional arguments, it's better to make sure that decorators accept both calling styles.

Anyway, I realize now that didn't push the very latest local version to the PR (only a couple minor changes) and this is actually done better now. If you want to take a look now to avoid losing context:

```
def register_factory(
    factory_class: Optional[factory.Factory] = None,
    *,
    model_fixture: Optional[str] = None,
    collection: Optional[MutableSequence[factory.Factory]] = None,
) -> Union[factory.Factory, Callable[[factory.Factory], factory.Factory]]:
    def _decorator(factory: factory.Factory) -> factory.Factory:
        if collection is not None:
            collection.append(factory)
        return pytfboy.register(factory, model_fixture)

    return _decorator(factory_class) if factory_class is not None else _decorator
```

Reply:
Thanks that helped! Got it! (after playing with it...)